### DISCUSSION TITLE
REST API and EJB APIs

### DISCUSSION THREAD

**Discussion by fco-code-org** [Keyfactor/ejbca-ce#535] [General]
Hello,

We are using the following EJBCA EJB APIs even if it is not recommended:
ejb:ejbca/ejbca-ejb**/EndEntityManagementSessionBean**
	**addUser**
	**deleteUser**
	**setUserStatus**
	**changeUser**

ejb:ejbca/ejbca-ejb/**EndEntityProfileSessionBean**
         **getEndEntityProfileId(String)**
         **getEndEntityProfile(String or int)**
         **getEndEntityProfileName**

ejb:ejbca/ejbca-ejb/**EndEntityAccessSessionBean**
          **findUser**
 
Does EJBCA has the equivalent of these above APIs in the REST APIs ?
If yes are these equivalent REST APIs available in EJBCA CE (Community Edition) ? 

Best Regards.

**Comment by primetomas** [comment]
You can find documentation on the different REST API endpoints here:
https://doc.primekey.com/ejbca/ejbca-operations/ejbca-ca-concept-guide/protocols/ejbca-rest-interface
And using the swagger API if you have a local dev install you can play with it.

End entity methods should be available in both Community and Enterprise. 
While for profile management the ConfigDump feature to manage profiles using yaml, is an Enterprise feature.

**Reply by fco-code-org** [reply]
Hello,

Looking at the URL https://doc.primekey.com/ejbca/ejbca-operations/ejbca-ca-concept-guide/protocols/ejbca-rest-interface#EJBCARESTInterface-Resourcesresources   I see that:
The /endentity/ resource handles end entity management. is an **ENTERPRISE** feature.

So I am lost because you said that   

> End entity methods should be available in both Community and Enterprise.

Doc issue ?

Best Regards.

**Reply by primetomas** [reply]
Sorry, it seems you are right and I am wrong. End entity is enterprise for now.

**Reply by fco-code-org** [reply]
Hello,

So it means that if we request EJBCA to generate leaf certitificates which requires end entity creation it is not possible with the REST API available in the EJBCA CE (Community Edition). Do you confirm ?

Is the EJBCA Soap APIs fully supported ? In such case do we have the same restriction for End Entity SOAP APIs if any ?
Is there a public documentation for the SOAP APIs as you provided for REST APIs ?

Could you confirm to me that all the REST and SOAP APIs are backward compatible ?

Best Regards.

**Reply by primetomas** [reply]
You can do this with the SOAP API yes. No restrictions. 
https://doc.primekey.com/ejbca/ejbca-operations/ejbca-ca-concept-guide/protocols/web-service-interface

APIs are backwards compatible. We have clients using SOAP API for 15 years without upgrade issues. REST API will also always be backwards compatible.

**Reply by fco-code-org** [reply]
Hello,

Great for backward compatibility :-) on both APIs (REST and SOAP).

Any reason for the End Entity REST APIs to be available only in the Enterprise  edition ?

Best Regards.

**Reply by pablo** [reply]
It's probably a business related reason. They need to monetize this product and this is a way of doing it.
